### ISSUE TITLE
Call return() on the iterator when a for-of loop exits abruptly

### ISSUE THREAD

**Issue by shicks** [google/closure-compiler#2958]
This breaks out one part of #2899.

I looked into this briefly today.  In order to do it right, we end up bloating our for-of transpilation quite a bit.  Currently we transpile

```js
for (var x of y) {
  ...
}
```

to

```js
for (var iter=$jscomp.makeIterator(y), key=iter.next(); !key.done; key=iter.next()) {
  let x = key.value;
  ...
}
```

which isn't so bad.  We can handle `break`, `continue`, and `return` pretty easily by inserting the `return` call immediately before each, and it doesn't cost anything if you don't use them.  But `throw` can be invisible (i.e. it's likely not coming from a `THROW` node in the block's AST), so the only way to handle it correctly is to wrap the whole thing in a catch:

```js
for (var iter=$jscomp.makeIterator(y), key=iter.next(); !key.done; key=iter.next()) {
  let x = key.value;
  try {
    ...
  } catch (e) {
    if (iter.return) iter.return();
    throw e;
  }
}
```

Even this only gets most of the way there - if `return` throws then we end up with the wrong error thrown.  To get that correct, we need an additional `try-finally` around the `return`.  If we're dealing with `break`s as well, at this point, it probably makes sense to handle those at the same time, rather than mutating the AST inside the body:

```js
try {
  for (var iter=$jscomp.makeIterator(y), key=iter.next(); !key.done; key=iter.next()) {
    let x = key.value;
    ...
  }
} catch (e) {
  key.thrown = {thrown: e};
} finally {
  try {
    if (!key.done && iter.return) iter.return();
  } finally {
    if (key.thrown) throw key.thrown.thrown;
  }
}
```

I estimate that this will add somewhere between 50 and 80 extra gzipped bytes to every for-of loop.

**Comment by shicks** [comment]
If we wrap the iterator in a wrapper that swallows any exception, we can get it down to

```js
try {
  for (var iter=$jscomp.makeIterator(y), key=iter.next(); !key.done; key=iter.next()) {
    let x = key.value;
    ...
  }
} finally {
  iter.close();
}
```

with a little extra runtime library.  This only amounts to 10-15 extra gzipped bytes, which is a lot better.

**Comment by trxcllnt** [comment]
Any update on this issue? Seems related to https://github.com/google/closure-compiler/issues/3971.

Here's is a simple repro I used to verify:

```shell
$ npx google-closure-compiler --js test.js --js_output_file test.min.js --language_in ECMASCRIPT_NEXT --language_out ECMASCRIPT5 --assume_function_wrapper --compilation_level ADVANCED --third_party true --formatting PRETTY_PRINT --debug
$ node test.js 
> return called: true
$ node test.min.js 
> return called: false
```

<details><summary>test.js:</summary><pre>
function first(xs) {<br/>    for (let x of xs) {<br/>        return x;<br/>    }<br/>}<br/><br/>let count = 0;<br/>let returnCalled = false;<br/><br/>const xs = {<br/>    [Symbol.iterator]() { return xs; },<br/>    next() {<br/>        if (count < 3) {<br/>            return { value: count++, done: false };<br/>        }<br/>        return { done: true };<br/>    },<br/>    return() {<br/>        returnCalled = true;<br/>        return { done: true };<br/>    }<br/>};<br/><br/>first(xs);<br/><br/>console.log('return called:', returnCalled);<br/></pre></details>

<details><summary>test.min.js:</summary><pre>function $$jscomp$arrayIteratorImpl$$($array$jscomp$6$$) {<br/>  var $index$jscomp$101$$ = 0;<br/>  return function() {<br/>    return $index$jscomp$101$$ < $array$jscomp$6$$.length ? {done:!1, value:$array$jscomp$6$$[$index$jscomp$101$$++]} : {done:!0};<br/>  };<br/>}<br/>var $$jscomp$defineProperty$$ = "function" == typeof Object.defineProperties ? Object.defineProperty : function($target$jscomp$92$$, $property$jscomp$5$$, $descriptor$jscomp$1$$) {<br/>  if ($target$jscomp$92$$ == Array.prototype || $target$jscomp$92$$ == Object.prototype) {<br/>    return $target$jscomp$92$$;<br/>  }<br/>  $target$jscomp$92$$[$property$jscomp$5$$] = $descriptor$jscomp$1$$.value;<br/>  return $target$jscomp$92$$;<br/>};<br/>function $$jscomp$getGlobal$$($passedInThis_possibleGlobals$$) {<br/>  $passedInThis_possibleGlobals$$ = ["object" == typeof globalThis && globalThis, $passedInThis_possibleGlobals$$, "object" == typeof window && window, "object" == typeof self && self, "object" == typeof global && global];<br/>  for (var $i$jscomp$3$$ = 0; $i$jscomp$3$$ < $passedInThis_possibleGlobals$$.length; ++$i$jscomp$3$$) {<br/>    var $maybeGlobal$$ = $passedInThis_possibleGlobals$$[$i$jscomp$3$$];<br/>    if ($maybeGlobal$$ && $maybeGlobal$$.Math == Math) {<br/>      return $maybeGlobal$$;<br/>    }<br/>  }<br/>  throw Error("Cannot find global object");<br/>}<br/>var $$jscomp$global$$ = $$jscomp$getGlobal$$(this);<br/>function $$jscomp$polyfill$$($property$jscomp$inline_11_split$jscomp$inline_8_target$jscomp$94$$, $impl$jscomp$inline_13_polyfill$jscomp$1$$) {<br/>  if ($impl$jscomp$inline_13_polyfill$jscomp$1$$) {<br/>    a: {<br/>      var $obj$jscomp$inline_7$$ = $$jscomp$global$$;<br/>      $property$jscomp$inline_11_split$jscomp$inline_8_target$jscomp$94$$ = $property$jscomp$inline_11_split$jscomp$inline_8_target$jscomp$94$$.split(".");<br/>      for (var $i$jscomp$inline_9_orig$jscomp$inline_12$$ = 0; $i$jscomp$inline_9_orig$jscomp$inline_12$$ < $property$jscomp$inline_11_split$jscomp$inline_8_target$jscomp$94$$.length - 1; $i$jscomp$inline_9_orig$jscomp$inline_12$$++) {<br/>        var $key$jscomp$inline_10$$ = $property$jscomp$inline_11_split$jscomp$inline_8_target$jscomp$94$$[$i$jscomp$inline_9_orig$jscomp$inline_12$$];<br/>        if (!($key$jscomp$inline_10$$ in $obj$jscomp$inline_7$$)) {<br/>          break a;<br/>        }<br/>        $obj$jscomp$inline_7$$ = $obj$jscomp$inline_7$$[$key$jscomp$inline_10$$];<br/>      }<br/>      $property$jscomp$inline_11_split$jscomp$inline_8_target$jscomp$94$$ = $property$jscomp$inline_11_split$jscomp$inline_8_target$jscomp$94$$[$property$jscomp$inline_11_split$jscomp$inline_8_target$jscomp$94$$.length - 1];<br/>      $i$jscomp$inline_9_orig$jscomp$inline_12$$ = $obj$jscomp$inline_7$$[$property$jscomp$inline_11_split$jscomp$inline_8_target$jscomp$94$$];<br/>      $impl$jscomp$inline_13_polyfill$jscomp$1$$ = $impl$jscomp$inline_13_polyfill$jscomp$1$$($i$jscomp$inline_9_orig$jscomp$inline_12$$);<br/>      $impl$jscomp$inline_13_polyfill$jscomp$1$$ != $i$jscomp$inline_9_orig$jscomp$inline_12$$ && null != $impl$jscomp$inline_13_polyfill$jscomp$1$$ && $$jscomp$defineProperty$$($obj$jscomp$inline_7$$, $property$jscomp$inline_11_split$jscomp$inline_8_target$jscomp$94$$, {configurable:!0, writable:!0, value:$impl$jscomp$inline_13_polyfill$jscomp$1$$});<br/>    }<br/>  }<br/>}<br/>$$jscomp$polyfill$$("Symbol", function($orig$jscomp$1$$) {<br/>  function $symbolPolyfill$$($opt_description$jscomp$2$$) {<br/>    if (this instanceof $symbolPolyfill$$) {<br/>      throw new TypeError("Symbol is not a constructor");<br/>    }<br/>    return new $SymbolClass$$($SYMBOL_PREFIX$$ + ($opt_description$jscomp$2$$ || "") + "_" + $counter$$++, $opt_description$jscomp$2$$);<br/>  }<br/>  function $SymbolClass$$($id$jscomp$5$$, $opt_description$jscomp$1$$) {<br/>    this.$g$ = $id$jscomp$5$$;<br/>    $$jscomp$defineProperty$$(this, "description", {configurable:!0, writable:!0, value:$opt_description$jscomp$1$$});<br/>  }<br/>  if ($orig$jscomp$1$$) {<br/>    return $orig$jscomp$1$$;<br/>  }<br/>  $SymbolClass$$.prototype.toString = function() {<br/>    return this.$g$;<br/>  };<br/>  var $SYMBOL_PREFIX$$ = "jscomp_symbol_" + (1E9 * Math.random() >>> 0) + "_", $counter$$ = 0;<br/>  return $symbolPolyfill$$;<br/>});<br/>$$jscomp$polyfill$$("Symbol.iterator", function($orig$jscomp$2_symbolIterator$$) {<br/>  if ($orig$jscomp$2_symbolIterator$$) {<br/>    return $orig$jscomp$2_symbolIterator$$;<br/>  }<br/>  $orig$jscomp$2_symbolIterator$$ = Symbol("Symbol.iterator");<br/>  for (var $arrayLikes$$ = "Array Int8Array Uint8Array Uint8ClampedArray Int16Array Uint16Array Int32Array Uint32Array Float32Array Float64Array".split(" "), $i$jscomp$6$$ = 0; $i$jscomp$6$$ < $arrayLikes$$.length; $i$jscomp$6$$++) {<br/>    var $ArrayLikeCtor$$ = $$jscomp$global$$[$arrayLikes$$[$i$jscomp$6$$]];<br/>    "function" === typeof $ArrayLikeCtor$$ && "function" != typeof $ArrayLikeCtor$$.prototype[$orig$jscomp$2_symbolIterator$$] && $$jscomp$defineProperty$$($ArrayLikeCtor$$.prototype, $orig$jscomp$2_symbolIterator$$, {configurable:!0, writable:!0, value:function() {<br/>      return $$jscomp$iteratorPrototype$$($$jscomp$arrayIteratorImpl$$(this));<br/>    }});<br/>  }<br/>  return $orig$jscomp$2_symbolIterator$$;<br/>});<br/>function $$jscomp$iteratorPrototype$$($iterator$jscomp$6_next$$) {<br/>  $iterator$jscomp$6_next$$ = {next:$iterator$jscomp$6_next$$};<br/>  $iterator$jscomp$6_next$$[Symbol.iterator] = function() {<br/>    return this;<br/>  };<br/>  return $iterator$jscomp$6_next$$;<br/>}<br/>var $count$$ = 0, $returnCalled$$ = !1, $$jscomp$compprop0$$ = {}, $xs$$ = ($$jscomp$compprop0$$[Symbol.iterator] = function() {<br/>  return $xs$$;<br/>}, $$jscomp$compprop0$$.next = function() {<br/>  return 3 > $count$$ ? {value:$count$$++, done:!1} : {done:!0};<br/>}, $$jscomp$compprop0$$.return = function() {<br/>  $returnCalled$$ = !0;<br/>  return {done:!0};<br/>}, $$jscomp$compprop0$$);<br/>a: {<br/>  var $JSCompiler_inline_result$jscomp$18$$;<br/>  var $iteratorFunction$jscomp$inline_20$$ = "undefined" != typeof Symbol && Symbol.iterator && $xs$$[Symbol.iterator];<br/>  if ($iteratorFunction$jscomp$inline_20$$) {<br/>    $JSCompiler_inline_result$jscomp$18$$ = $iteratorFunction$jscomp$inline_20$$.call($xs$$);<br/>  } else {<br/>    if ("number" == typeof $xs$$.length) {<br/>      $JSCompiler_inline_result$jscomp$18$$ = {next:$$jscomp$arrayIteratorImpl$$($xs$$)};<br/>    } else {<br/>      throw Error(String($xs$$) + " is not an iterable or ArrayLike");<br/>    }<br/>  }<br/>  for (var $$jscomp$key$m1422502875$0$x$jscomp$inline_16$$ = $JSCompiler_inline_result$jscomp$18$$.next(); !$$jscomp$key$m1422502875$0$x$jscomp$inline_16$$.done;) {<br/>    break a;<br/>  }<br/>}<br/>console.log("return called:", $returnCalled$$);<br/><br/></pre></details>
